File: 6th_Sem/CSS651_Compiler_Lab/Lab_7/q16.py
class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None
        self.rule = None

def postfix_expr(root):
    if root.left is None and root.right is None:
        return root.value
    else:
        left = postfix_expr(root.left)
        right = postfix_expr(root.right)
        return left + " " + right + " " + root.rule

def parse_expr(expr):
    stack = []
    for char in expr:
        if char.isdigit():
            node = Node(char)
            node.rule = ". → '" + char + "'"
            stack.append(node)
        else:
            if len(stack) < 2:
                print("Error: Invalid input expression")
                return None
            right = stack.pop()
            left = stack.pop()
            node = Node(char)
            node.rule = char + " → ."
            node.left = left
            node.right = right
            stack.append(node)
    if len(stack) != 1:
        print("Error: Invalid input expression")
        return None
    return stack[0]

expr = "9-4+1"

root = parse_expr(expr)

if root is not None:
    postfix = postfix_expr(root)

    print("Postfix expression:", postfix)

    def print_tree(node, level=0):
        if node is not None:
            print("  " * level + "+- " + str(node.rule))
            print_tree(node.left, level + 1)
            print_tree(node.right, level + 1)

    print("Annotated parse tree:")
    print_tree(root)

File: 6th_Sem/CSS651_Compiler_Lab/Lab_7/q16.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_SIZE 100

typedef struct Node {
    char* data;
    struct Node* left;
    struct Node* right;
} Node;

typedef struct Stack {
    int top;
    char* data[MAX_SIZE];
} Stack;

Node* makeNode(char* data, Node* left, Node* right) {
    Node* node = (Node*)malloc(sizeof(Node));
    node->data = data;
    node->left = left;
    node->right = right;
    return node;
}

Stack* makeStack() {
    Stack* stack = (Stack*)malloc(sizeof(Stack));
    stack->top = -1;
    return stack;
}

int isEmpty(Stack* stack) {
    return stack->top == -1;
}

void push(Stack* stack, char* data) {
    stack->data[++stack->top] = data;
}

char* pop(Stack* stack) {
    if (isEmpty(stack)) {
        printf("Stack is empty!\n");
        exit(EXIT_FAILURE);
    }
    return stack->data[stack->top--];
}

char* peek(Stack* stack) {
    if (isEmpty(stack)) {
        printf("Stack is empty!\n");
        exit(EXIT_FAILURE);
    }
    return stack->data[stack->top];
}

int isOperand(char symbol) {
    return symbol >= '0' && symbol <= '9';
}

int precedence(char operator) {
    switch (operator) {
        case '+':
        case '-':
            return 1;
        case '*':
        case '/':
            return 2;
        default:
            printf("Invalid operator!\n");
            exit(EXIT_FAILURE);
    }
}

void infixToPostfix(char* infix, char* postfix) {
    Stack* stack = makeStack();
    int i, j;
    for (i = 0, j = 0; infix[i]; i++) {
        if (isOperand(infix[i])) {
            postfix[j++] = infix[i];
        }
        else if (infix[i] == '+' || infix[i] == '-' || infix[i] == '*' || infix[i] == '/') {
            while (!isEmpty(stack) && precedence(peek(stack)[0]) >= precedence(infix[i])) {
                postfix[j++] = ' ';
                postfix[j++] = pop(stack)[0];
            }
            postfix[j++] = ' ';
            char* op = (char*)malloc(sizeof(char) * 2);
            op[0] = infix[i];
            op[1] = '\0';
            push(stack, op);
        }
    }
    while (!isEmpty(stack)) {
        postfix[j++] = ' ';
        postfix[j++] = pop(stack)[0];
    }
    postfix[j] = '\0';
}

void inorder(Node* node) {
    if (node) {
        inorder(node->left);
        printf("%s", node->data);
        inorder(node->right);
    }
}

int main() {
    char infix[] = "9-4+1";
    char postfix[MAX_SIZE];
    infixToPostfix(infix, postfix);
    printf("Infix expression: %s\n", infix);
    printf("Postfix expression: %s\n", postfix);

    Node* root = makeNode("+", makeNode("-", makeNode("9", NULL, NULL), makeNode("4", NULL, NULL)), makeNode("1", NULL, NULL));
    printf("Annotated parse tree: ");
    inorder(root);
    printf("\n");

    return 0;
}
